util/logutil/consistency/reporter.go

```diff
@@ -0,0 +1,264 @@
+	if respValue.Info != nil {
+		var (
+			hd    kv.Handle
+			err   error
+			datas = make(map[string]map[string]string)
+		)
+		for _, w := range respValue.Info.Writes {
+			if len(w.ShortValue) > 0 {
+				hd, err = tablecodec.DecodeIndexHandle(key, w.ShortValue, len(r.Idx.Columns))
+				if err == nil {
+					datas[strconv.FormatUint(w.StartTs, 10)] = map[string]string{"handle": hd.String()}
+				}
+			}
+		}
+		for _, v := range respValue.Info.Values {
+			if len(v.Value) > 0 {
+				hd, err = tablecodec.DecodeIndexHandle(key, v.Value, len(r.Idx.Columns))
+				if err == nil {
+					datas[strconv.FormatUint(v.StartTs, 10)] = map[string]string{"handle": hd.String()}
+				}
+			}
+		}
+		if len(datas) > 0 {
+			outMap["decoded"] = datas
+			if err != nil {
+				outMap["decode_error"] = err.Error()
+			}
+		}
+	}
+}
+
+func decodeMvccData(bs []byte, colMap map[int64]*types.FieldType, tb *model.TableInfo) (map[string]string, error) {
+	rs, err := tablecodec.DecodeRowToDatumMap(bs, colMap, time.UTC)
+	record := make(map[string]string, len(tb.Columns))
+	for _, col := range tb.Columns {
+		if c, ok := rs[col.ID]; ok {
+			data := "nil"
+			if !c.IsNull() {
+				data, err = c.ToString()
+			}
+			record[col.Name.O] = data
+		}
+	}
+	return record, err
+}
+
+// ReportLookupInconsistent reports inconsistent when index rows is more than record rows.
+func (r *Reporter) ReportLookupInconsistent(ctx context.Context, idxCnt, tblCnt int, missHd, fullHd []kv.Handle, missRowIdx []RecordData) error {
+	if r.Sctx.GetSessionVars().EnableRedactLog {
+		logutil.Logger(ctx).Error("indexLookup found data inconsistency",
+			zap.String("table_name", r.Tbl.Name.O),
+			zap.String("index", r.Idx.Name.O),
+			zap.Int("index_cnt", idxCnt),
+			zap.Int("table_cnt", tblCnt))
+	} else {
+		fs := []zap.Field{
+			zap.String("table_name", r.Tbl.Name.O),
+			zap.String("index_name", r.Idx.Name.O),
+			zap.Int("index_cnt", idxCnt), zap.Int("table_cnt", tblCnt),
+			zap.String("missing_handles", fmt.Sprint(missHd)),
+			zap.String("total_handles", fmt.Sprint(fullHd)),
+		}
+		for i, hd := range missHd {
+			fs = append(fs, zap.String("row_mvcc_"+strconv.Itoa(i), getMvccByKey(r.Sctx, r.HandleEncode(hd), r.decodeRowMvccData)))
+		}
+		for i, rowIdx := range missRowIdx {
+			fs = append(fs, zap.String("index_mvcc_"+strconv.Itoa(i), getMvccByKey(r.Sctx, r.IndexEncode(&rowIdx), r.decodeIndexMvccData)))
+		}
+		logutil.Logger(ctx).Error("indexLookup found data inconsistency", fs...)
+	}
+	return ErrLookupInconsistent.GenWithStackByArgs(r.Tbl.Name.O, r.Idx.Name.O, idxCnt, tblCnt)
+}
+
+// ReportAdminCheckInconsistentWithColInfo reports inconsistent when the value of index row is different from record row.
+func (r *Reporter) ReportAdminCheckInconsistentWithColInfo(ctx context.Context, handle kv.Handle, colName string, idxDat, tblDat fmt.Stringer, err error, idxRow *RecordData) error {
```

Comment:
🤔 Why `idxDat, tblDat fmt.Stringer` accepts Stringer instead of Datum?

Reply:
this method only uses the function of `fmt.Stringer`, in some middle versions it want to accept more data type like rowRecord...but now its only Datum..

and this also make code more simple,  we want to call datum's String() in fmt.Sprinter, so we need `*Datum` instead of `Datum`, but GetDatum() always return a `Datum`, so if we ReportAdminCheckInconsistentWithColInfo's param to `*Datum` it need alloc additional variable and pass point to method in caller...if we let param be `Datum`, we need `fmt.Sprint(&a)` and alloc more for Datum copy?